src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionService.cs

```diff
@@ -0,0 +1,667 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+using EditorCompletion = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using RoslynTrigger = Microsoft.CodeAnalysis.Completion.CompletionTrigger;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.EditorImplementation
+{
+    internal class EditorAsyncCompletionService : IAsyncCompletionItemManager
+    {
+        private readonly IAsyncCompletionBroker _broker;
+        private readonly CompletionHelper _completionHelper;
+
+        private const int MaxMRUSize = 10;
+        private ImmutableArray<string> _recentItems = ImmutableArray<string>.Empty;
+
+        public EditorAsyncCompletionService(IAsyncCompletionBroker broker)
+        {
+            _broker = broker;
+            _completionHelper = new CompletionHelper(isCaseSensitive: true);
+        }
+
+        public Task<ImmutableArray<EditorCompletion.CompletionItem>> SortCompletionListAsync(
+            IAsyncCompletionSession session,
+            EditorCompletion.AsyncCompletionSessionInitialDataSnapshot data,
+            CancellationToken token)
+        {
+            session.ItemCommitted += ItemCommitted;
+            session.Dismissed += SessionDismissed;
+            return Task.FromResult(data.InitialList.OrderBy(i => i.SortText).ToImmutableArray());
+        }
+
+        private void ItemCommitted(object sender, EditorCompletion.CompletionItemEventArgs e)
+        {
+            MakeMostRecentItem(e.Item.DisplayText);
+        }
+
+        private void SessionDismissed(object sender, EventArgs e)
+        {
+            // TODO: Unhook the session's events when the session is available in the args
```

Comment:
session is the `sender`. 
Note to self: check if documentation explains that `sender` is `IAsyncCompletionSession`